lighthouse-core/scripts/pptr-run-devtools.js

```diff
@@ -0,0 +1,165 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/**
+ * USAGE:
+ * URL list file: yarn run-devtools < path/to/urls.txt
+ * Single URL: yarn run-devtools "https://example.com"
+ */
+
+const puppeteer = require('puppeteer');
+const fs = require('fs');
+const readline = require('readline');
+const glob = require('glob');
+
+/** @typedef {{result?: {value?: string, objectId?: number}, exceptionDetails?: object}} RuntimeEvaluateResponse */
+
+const OUTPUT_DIR = 'latest-run/devtools-lhrs';
+
+const urlArg = process.argv[2];
+
+/**
+ * https://source.chromium.org/chromium/chromium/src/+/master:third_party/devtools-frontend/src/front_end/test_runner/TestRunner.js;l=170;drc=f59e6de269f4f50bca824f8ca678d5906c7d3dc8
+ * @param {Record<string, function>} receiver
+ * @param {string} methodName
+ * @param {function} override
+ */
+function addSniffer(receiver, methodName, override) {
+  const original = receiver[methodName];
+  if (typeof original !== 'function') {
+    throw new Error('Cannot find method to override: ' + methodName);
+  }
+
+  /**
+   * @param  {...any} args
+   */
+  receiver[methodName] = function(...args) {
+    let result;
+    try {
+      result = original.apply(this, args);
+    } finally {
+      receiver[methodName] = original;
+    }
+    // In case of exception the override won't be called.
+    try {
+      Array.prototype.push.call(args, result);
+      override.apply(this, args);
+    } catch (e) {
+      throw new Error('Exception in overriden method \'' + methodName + '\': ' + e);
+    }
+    return result;
+  };
+}
+
+const sniffLhr = `
+new Promise(resolve => {
+  (${addSniffer.toString()})(
+    UI.panels.lighthouse.__proto__,
+    '_buildReportUI',
+    (lhr, artifacts) => resolve(lhr)
+  );
+});
+`;
+
+const startLighthouse = `
+(() => {
+  UI.ViewManager.instance().showView('lighthouse');
```

Comment:
is this working for you locally still? When I run it it seems like this line isn't succeeding (it's stuck on the elements panel)

Reply:
It's working fine for me on the latest Canary, but I was having issues earlier with a custom DT frontend.

Can you try opening DT on DT and running this line manually?

Reply:
Maybe we need to add an extra `ViewManager`?

https://source.chromium.org/chromium/chromium/src/+/master:third_party/devtools-frontend/src/front_end/test_runner/TestRunner.js;l=264-270?q=TestRunner.js&ss=chromium%2Fchromium%2Fsrc

Reply:
Can you do `const ViewManager = UI.ViewManager.ViewManager || UI.ViewManager; ...` so it works in stable and canary?

Reply:
Update: The issue here was that `CHROME_PATH` wasn't set, so puppeteer defaulted to M78 which uses the "Audits" panel instead of the "Lighthouse" panel.

Reply:
I've upgraded our puppeteer dep: #12284